openwisp_controller/config/whois/tasks.py

```diff
@@ -0,0 +1,133 @@
+        return super().on_failure(exc, task_id, args, kwargs, einfo)
+
+
+# device_pk is used when task fails to report for which device failure occurred
+@shared_task(
+    bind=True,
+    base=WhoIsCeleryRetryTask,
+    **app_settings.API_TASK_RETRY_OPTIONS,
+)
+def fetch_whois_details(self, device_pk, initial_ip_address, new_ip_address):
+    """
+    Fetches the WhoIs details of the given IP address
+    and creates/updates the WhoIs record.
+    """
+    WhoIsInfo = load_model("config", "WhoIsInfo")
+
+    try:
+        # Host is based on the db that is used to fetch the details.
+        # As we are using GeoLite2, 'geolite.info' host is used.
+        # Reference: https://geoip2.readthedocs.io/en/latest/#sync-web-service-example
+        ip_client = geoip2_webservice.Client(
+            account_id=app_settings.GEOIP_ACCOUNT_ID,
+            license_key=app_settings.GEOIP_LICENSE_KEY,
+            host="geolite.info",
+        )
+
+        data = ip_client.city(ip_address=new_ip_address)
+        # Format address using the data from the geoip2 response
+        address = {
+            "city": getattr(data.city, "name", ""),
+            "country": getattr(data.country, "name", ""),
+            "continent": getattr(data.continent, "name", ""),
+            "postal": str(getattr(data.postal, "code", "")),
+        }
+        # Create the WhoIs information
+        WhoIsInfo.objects.create(
+            organization_name=data.traits.autonomous_system_organization,
+            asn=data.traits.autonomous_system_number,
+            country=data.country.name,
+            timezone=data.location.time_zone,
+            address=address,
+            cidr=data.traits.network,
+            ip_address=new_ip_address,
+        )
+        logger.info(f"Successfully fetched WHOIS details for {new_ip_address}.")
+
+        # not using the delete method of WhoIsService due to circular imports
+        queryset = WhoIsInfo.objects.filter(ip_address=initial_ip_address)
+        # the following check ensures that for a case when device last_ip
+        # is not changed and there is no related whois record, we do not
+        # delete the newly created record as both `initial_ip_address` and
+        # `new_ip_address` would be same for such case. Also invalidate the cache
+        if initial_ip_address != new_ip_address and queryset.exists():
+            # If any active devices are linked to the following record,
+            # then they will trigger this task and new record gets created
+            # with latest data.
+            queryset.delete()
+            cache.delete(f"whois_info:{initial_ip_address}")
+
+    # Catching all possible exceptions raised by the geoip2 client
+    # logging the exceptions and raising them with appropriate messages
+    except errors.AddressNotFoundError:
+        message = _(f"No WHOIS information found for IP address {new_ip_address}.")
```

Comment:
These messages can be stored in a dictionary with the exception classes as keys.